packages/engine/lib/simulation-control/src/engine.rs

```diff
@@ -274,6 +274,11 @@ impl Engine {
     fn handle_messages(&mut self, state: &mut State, message_map: &MessageMap) -> Result<()> {
         let message_proxies = state.message_pool().read_proxies()?;
         let mut commands = Commands::from_hash_messages(message_map, &message_proxies)?;
+        // it is necessary to drop `message_proxies` here, because it contains a strong reference to
+        // the MessageBatch; if this strong reference exists, then it is not possible to call
+        // `state.read()` later (because `state.read()` will fail it there are multiple references
+        // to the batch).
+        drop(message_proxies);
```

Comment:
Can this be rewritten or will the borrow checker complain?
```suggestion
        let mut commands = {
            // it is necessary to drop `message_proxies` after reading the commands because it contains a strong reference to
            // the `MessageBatch`; if this strong reference exists, then it is not possible to call
            // `state.read()` later (because `state.read()` will fail it there are multiple references
            // to the batch).
            let message_proxies = state.message_pool().read_proxies()?;
            Commands::from_hash_messages(message_map, &message_proxies)
        };
```

(Will most probably mess up rustfmt)

Reply:
Awesome!
Yes, I missed the `?` 😄